Implement metadata views

maggo opened the pull request:
This adds metadata views as required by the NFT catalogue.

Based on #1 which should be merged first

---

Supported views:

```
MetadataViews.Display
MetadataViews.Editions
MetadataViews.NFTCollectionDisplay
MetadataViews.ExternalURL
MetadataViews.NFTCollectionData
MetadataViews.Royalties
MetadataViews.Serial
```

## Todo

* [x] Mapping and fallback for legacy template thumbnails
* [x] Implement dynamic editions functionality (Editions based nft type)
* [x] Set website URL fallback
* [ ] Add functionality to generate dynamic socials list
* [x] Implement royalties

maggo committed:
Add first draft MetadataViews integration and script to get the generated NFT view

maggo committed:
Add instagram social link

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -228,6 +231,98 @@ pub contract XGStudio: NonFungibleToken {
+                    ])
+                case Type<MetadataViews.NFTCollectionDisplay>():
+                    return MetadataViews.NFTCollectionDisplay(
+                        name: brand.brandName == "xGrunning" ? "xGMove" : brand.brandName,
```
I've updated the brand metadata with name field, we should replace the line with 
```suggestion
                        name: brand.brandName == brand.data["name"],
```

maggo replied:
Done! Since it's always set we can simplify to `name: brand.data["name"] ?? "",`

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -228,6 +231,98 @@ pub contract XGStudio: NonFungibleToken {
+                    return MetadataViews.NFTCollectionDisplay(
+                        name: brand.brandName == "xGrunning" ? "xGMove" : brand.brandName,
+                        description: brand.data["description"] ?? "",
+                        externalURL: MetadataViews.ExternalURL(brand.data["websiteUrl"] ?? "https://xgrunning.io"), // @TODO: Update fallback
```
It really should never happen as we need this for the the external url in the Display view but let's have

```suggestion
                        externalURL: MetadataViews.ExternalURL(brand.data["websiteUrl"] ?? "https://xgstudios.io"), // @TODO: Update fallback
```

maggo replied:
Done!

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -228,6 +231,98 @@ pub contract XGStudio: NonFungibleToken {
+                            mediaType: "image/png"
+                        ),
+                        socials: {
+                            // @TODO: Implement dynamic socials list
```
would it make sense to add `social` field on brand.data containing a string with comma separated list of available socials which then would be parsed into an array and iterated over like:
`
brand.data = {
              ....,
             "socials": "twitter,instagram,discord",
             "twitter": "url1",
             "instagram": "url2",
              "discord": "url3"
              ...,
  }
  
  // here the magic happens ;) meaning iI don't know if possible
  let availableSocials = split(brand.data["socials"], ",")
  for social in availableSocials {
       socials[social] = brand.data[social]
  }

`

maggo replied:
cadence doesn't have string.split :( We'd need to find another way to configure this. As discussed, updating the data field to be `{String: AnyStruct}` or adding a new field with this type might be an option so we have 
```
socials: {
  twitter: "https://twitter",
  discord: "https://discord"
}
```

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -228,6 +231,98 @@ pub contract XGStudio: NonFungibleToken {
+                        }
+                    )
+                case Type<MetadataViews.ExternalURL>():
+                    return MetadataViews.ExternalURL((brand.data["websiteUrl"] ?? "https://xgrunning.io").concat("/rewards/").concat(self.id.toString())) // @TODO: Update fallback
```
It won't work at first, but will implement that later on 

```suggestion
                    return MetadataViews.ExternalURL((brand.data["websiteUrl"] ?? "https://xgstudios.io").concat("/rewards/").concat(self.id.toString())) // @TODO: Update fallback
```

maggo committed:
Implement thumbnail mapping and support for IPFS files

maggo committed:
Update brand data keys for name and images

maggo committed:
Update fallback URL

maggo commented:
* Added thumbnail mapping
* Updated brand data keys and socials
* Updated fallback URL

## Todo
* Dynamic editions
* Royalties
* Dynamic Social links, dynamic brand data

maggo committed:
Implement dynamic editions functionality

maggo committed:
Add dummy royalties receiver for NFTViews compatibility

maggo committed:
Fix borrowXGStudio_NFT return type cast

maggo committed:
Add London Duathlon 2022 thumbnail

maggo committed:
Add Hat Trick football activityType

yellowBirdy committed:
getThumbnailFile small updates

- switch updated from `activityType` to `xGRewardType` getThumbnailFile
- standard thumbnail as CID from  "thumbnail" key on the templateData
- simplified the Hackney Half thumbs logic to resemble the ASICS one

maggo committed:
Remove royalties

maggo committed:
Implement Medias view

maggo committed:
Make tests compatible with Metadataviews extension

maggo committed:
Implement shallResolve and shallPass helpers for tests

maggo committed:
Implement NFTView test

maggo committed:
First pass receivers definition

yellowBirdy committed:
Descriptions for legacy football templates

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -476,6 +477,54 @@ pub contract XGStudio: NonFungibleToken {
+                MetadataViews.Royalty(
+                    receiver: genericReceiver,
+                    cut: 0.025,
+                    description: "xGStudios Cut"
```
`                    description: "Artist"
`

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -476,6 +477,54 @@ pub contract XGStudio: NonFungibleToken {
+                        MetadataViews.Royalty(
+                            receiver: receiver,
+                            cut: 0.05,
+                            description: "xGFootball Cut"
```
`                            description: "xGFootball treasury"
`

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -476,6 +477,54 @@ pub contract XGStudio: NonFungibleToken {
+                        MetadataViews.Royalty(
+                            receiver: receiver,
+                            cut: 0.05,
+                            description: "xGMove Cut"
```
`description: "xGMove treasury"
`

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -476,6 +477,54 @@ pub contract XGStudio: NonFungibleToken {
+                    )
+                }
+                
+            } else {
+                // @TODO: Identify xGMove events
```
`} else if  (tempalte.brandId == 2){`

yellowBirdy commented on contracts/XGStudio.cdc:
```diff
@@ -476,6 +477,54 @@ pub contract XGStudio: NonFungibleToken {
+                )
+            ] : []
+
+            if (templateData["activityType"] as! String? == "Football") {
```
} else if (tempalte.brandId == 1){

maggo committed:
Implement royalties feedback

maggo committed:
Switch brand based royalties receivers

yellowBirdy committed:
Added tiktok to socials

yellowBirdy committed:
Updated tests with tiktok

yellowBirdy committed:
fixing the thumbnail IPFSFile slash bug

maggo committed:
Set fallback values to make switch cases work
